src/tools/apollo-tool/apollo-tuner.cpp

```diff
@@ -0,0 +1,313 @@
+    return static_cast<float>(value.value.double_value);
+  case Kokkos::Tools::Experimental::ValueType::kokkos_value_string:
+    return static_cast<float>(
+        std::hash<std::string>{}(value.value.string_value)); // terrible
+  default:
+    return 0.0; // break in some fashion instead
+  }
+}
+
+Kokkos::Tools::Experimental::VariableValue mvv(size_t index, Kokkos::Tools::Experimental::VariableValue reference, Kokkos::Tools::Experimental::ValueSet& set){
+  Kokkos::Tools::Experimental::VariableValue value;
+  value.type_id = reference.type_id;
+  value.metadata = reference.metadata;
+  switch(reference.metadata->type){
+  case Kokkos::Tools::Experimental::ValueType::kokkos_value_int64:
+    value.value.int_value = set.values.int_value[index];  
+  break;
+  case Kokkos::Tools::Experimental::ValueType::kokkos_value_double:
+    value.value.double_value = set.values.double_value[index];  
+  break;
+  case Kokkos::Tools::Experimental::ValueType::kokkos_value_string:
+    strncpy(value.value.string_value,set.values.string_value[index],KOKKOS_TOOLS_TUNING_STRING_LENGTH);
+  break;
+  }
+  return value;
+}
+
+
+std::string get_region_name(variableSet variables){
+  std::string name;
+  for(int x =0 ; x< variables.num_variables; ++x){
+    auto* local_name = reinterpret_cast<std::string*>(variables.variable_ids[x].metadata->toolProvidedInfo);
+    name += *local_name;    
+    if( x == (variables.num_variables - 1)){
+      name+=".yaml";
+    }
+    else {
+      name+="_";
+    }
+  }
+  return name;
+}
+bool file_exists(std::string path){
+  struct stat buffer;   
+  return (stat (path.c_str(), &buffer) == 0); 
+}
+extern "C" void kokkosp_request_values(
+    size_t contextId, size_t numContextVariables,
+    Kokkos::Tools::Experimental::VariableValue *contextValues, size_t numTuningVariables,
+    Kokkos::Tools::Experimental::VariableValue *tuningValues
+    ) {
+  if (numTuningVariables == 0) {
+    return;
+  }
+  static int created_regions;
+  // std::cout << "Have "<<numContextVariables<<" context,
+  // "<<numTuningVariables<<" tuning\n";
+  int64_t choiceSpaceSize = 1;
+  for (int x = 0; x < numTuningVariables; ++x) {
+      if (choiceSpaceSize > max_choices / tuningValues[x].metadata->candidates.set.size) {
+        printf("Apollo Tuner: too many choices\n");
+      }
+      choiceSpaceSize *= tuningValues[x].metadata->candidates.set.size;
+  }
+  variableSet tuningProblem;
+  int numValidVariables = 0;
+  for (int x = 0; x < numContextVariables; ++x) {
+    if (untunables.find(contextValues[x].type_id) == untunables.end()) {
+      tuningProblem.variable_ids[numValidVariables++] = contextValues[x];
+    }
+  }
+  for (int x = 0; x < numTuningVariables; ++x) {
+      tuningProblem.variable_ids[numValidVariables++] = tuningValues[x];
+  }
+  tuningProblem.num_variables = numValidVariables;
+  auto iter = tuning_regions.emplace(tuningProblem, std::make_pair("",nullptr));
```

Comment:
I see you don't like this code!

Is this more aesthetically pleasing?
````cpp
auto iter = tuning_regions.find(tuning_problem);
if (iter == tuning_regions.end()){
  //bla bla bla
  iter = tuning_regions.emplace(...)
}
region = iter->second.second;
````

Reply:
Yeah, I'm not a huge fan of either, but I can't even pretend it's rational. I appreciate the suggestion, but I think what we have is decent, and appears to match standard practice